### ISSUE TITLE
Should Semimodules and Modules have a rule saying that left- and right-multiplication are the same?

### ISSUE THREAD

**Issue by Taneb** [agda/agda-stdlib#1888]
i.e. `∀ x m → x *ₗ m ≈ᴹ m *ᵣ x`.

This can't be stated for `Bisemimodule` and `Bimodule` but for `Semimodule` and `Module` it feels a useful property to have that we can't currently prove for a general module.

**Comment by MatthewDaggitt** [comment]
I have no idea about module theory, so I'm not really equipped to answer this one.

**Comment by jamesmckinna** [comment]
I've been doing some work recently with modules, and I think this property is *not* in general provable. But best place to ask: mathoverflow imho

**Comment by Taneb** [comment]
I've found an example of a module (by our definition) where the two operations are different.

Take G to be the usual commutative of Gaussian integers (complex numbers of the form x + i*y for x, y integers).

Then G acts as a left module over itself with multiplication.
We can also take G to be a right module over itself with the operation `\x y -> x * conjugate y`.

With these two definitions, we can build a module according to stdlib's understanding of the term. See here: https://gist.github.com/Taneb/64cd57ef1896a820b27921131cfab5a6

I claim this is strongly undesirable.

**Comment by JacquesCarette** [comment]
It took me a while to reconstruct what @Taneb found objectionable, so I'll expand it out in this comment: normally a module has a single notion of multiplication, not a left one and a right one. If one builds up module by joining left/right modules, then it's possible to create a module where they are different.

The fundamental issue here is that there is no simple syntactic embedding going on here. The relation between the theories involved are an almost-trivial theorem, but not a trivial theorem (i.e. the proof is not exactly 'refl').

The simplest fix is as the title says: it should be an axiom that left- and right- multiplication amount to the same thing. They are not quite 'the same' in the strong sense of 'the same', but they are trivially equivalent...

**Comment by jamesmckinna** [comment]
But see also the discussion [here](https://math.stackexchange.com/questions/1493609/right-module-vs-left-module), as well as [this](https://mathoverflow.net/questions/64370/simplest-examples-of-rings-that-are-not-isomorphic-to-their-opposites) which links from the first; also [this]()https://math.stackexchange.com/questions/535246/on-similarities-and-differences-between-right-and-left-modules-over-a-ring). 

Notwithstanding @Taneb 's wish to avoid a "strongly undesirable" state of affairs, I wonder if the designers of `Algebra.Module.Structures` might need to reconsider their definitions? I confess I don't know how to reconcile the examples given above with @JacquesCarette 's assertion that the left and right- actions are "trivially equivalent".

**Comment by JacquesCarette** [comment]
Upshot: it seems to make sense to define (R,S)-module, which is actually what is currently implemented. It then makes sense to define R-module only for commutative rings R. (Or, at least, I think that's the conclusion to draw from all those MO posts.)

**Comment by jamesmckinna** [comment]
Jacques, thanks for the sensitive editing. But I draw/drew the opposite conclusion to you! (Especially because the noncommutative case ought not to have to take a back seat for the sake of the common case). 

From my reading, the MO examples surely (?) make clear that the notion of R-S-bimodule is independently of interest... and that in the commutative case, one should prove the lemma that a left-module structure gives rise to a right module structure, and hence, if somewhat boringly, an R-R-binodule structure. 

The hard choice is how to bias the left/right choice when defining 'module'. But the articles are clear there, too, at least as I understand them: you actually have to choose, and similarly to the above, prove the lemma that a left R-module is (almost automatically) a right Rop-module.

**Comment by JacquesCarette** [comment]
Opposite? I too drew the conclusion that R-S-bimodules are of independent interest, and perhaps the most natural definition to arrive at 'structurally'. How to then also define reasonably convenient "traditional" R-modules remains open. My conclusion is that maybe it should be defined structurally.

**Comment by jamesmckinna** [comment]
Hmm, perhaps I see your point, but do you further intend that in the commutative R case
* the left- and right- multiplications should be made definitionally equal? and
* the notion of R-R-bimodule is retained, and remains strictly weaker? 

In the first, that still feels as though the user needs to commit to making a biased choice when defining any particular instance of such structure, even if you were to have (additional) smart constructors to do the identifications for you?

**Comment by jamesmckinna** [comment]
And the relevant definition in `Algebra.Module.Structures` *does* need to be changed: 
```agda
  -- An R-module is an R-R-bimodule where R is commutative.
  -- This means that *ₗ and *ᵣ coincide up to mathematical equality, though it
  -- may be that they do not coincide up to definitional equality.
```

**Comment by Taneb** [comment]
@JacquesCarette what do you mean by "defined structurally" here?

**Comment by JacquesCarette** [comment]
@Taneb  I meant that a definition that inherits from lower things, the way the 'hierarchy' in `Algebra` is currently done, in a straight syntactic manner.  By non-structural, I was thinking of a more-or-less direct definition of R-module, from which you could *reconstruct* an R-R-bimodule structure via a function rather than as a projector of fields.

@jamesmckinna Yes, I think R-R-bimodule seems to be a strictly weaker notion. And yes, I mean for the left- and right- multiplications to be definitionally equal by, in fact, only storing a single multiplication in the structure itself.

**Comment by jamesmckinna** [comment]
I guess I'll move further comments to the new issues and the associated PRs from now on.

**Comment by jamesmckinna** [comment]
A partial answer to the first of the above

> the construction R ᵒᵖ of the opposite ring of a Ring R (Semiring? Group? Monoid? Magma?);

would seem to be `Algebra.Construct.Flip.Op`... PR to follow.

**Comment by jamesmckinna** [comment]
Regarding

> the isomorphism proof is a property. `Algebra.Ring.Property` ?

I'd be happy with both the assertion, and the location. But I can't/couldn't help worrying about needing to capture the iso for all the (suitable) substructures of `Ring`, before finally realsing the last one as the lemma/proeprty that I wanted. So `Algebra.Ring` doesn't seem... *quite* right?

**Comment by jamesmckinna** [comment]
Apologies: It occurred to me that these discussions have again strayed from the main issue, and perhaps should be opened as a separate issue linking from PR #1900 ...?

**Comment by JacquesCarette** [comment]
You're right that this has drifted, but since I think the side discussion is also concluding, it doesn't seem worth opening a new one, just to have it end quickly.  So: Since it's a construction, it shouldn't go in `Algebra.Ring.Property`. The construction might well be long in the same place where `Module` itself is defined. Or in some `Algebra.Construction` module (probably better that).

If the iso is about structure `X`, and then it's re-used for `Ring`, maybe `Algebra.X.Property` ? There really will be properties specific to each rung of the algebra ladder.

**Comment by jamesmckinna** [comment]
Hmm. OK. But I think the constructions/iso can now wait until the PR is settled. But I have one more thing for my wishlist (or a subsequent PR after #1898... ;-)) namely that every (commutative)(semi)ring should also give rise to the appropriate module structure over *itself*: candidate module `Algebra.Module.Construct.Self`?

**Comment by Taneb** [comment]
> Hmm. OK. But I think the constructions/iso can now wait until the PR is settled. But I have one more thing for my wishlist (or a subsequent PR after #1898... ;-)) namely that every (commutative)(semi)ring should also give rise to the appropriate module structure over _itself_: candidate module `Algebra.Module.Construct.Self`?

@jamesmckinna if you mean what I think you mean, then that exists! Under the somewhat non-obvious name of [Algebra.Module.Construct.TensorUnit](https://agda.github.io/agda-stdlib/Algebra.Module.Construct.TensorUnit.html)

**Comment by jamesmckinna** [comment]
Yes, of course! Thanks very much! 
That will greatly simplify my work on `Algebra.Module.Construct.Nagata`

**Comment by MatthewDaggitt** [comment]
Yup I agree with @JacquesCarette on the placement issue. Constructions should go in the `Algebra.Construct` folder, while non-constructions or concrete instantiations of those constructions should go in `Algebra.X.Properties`.

**Comment by jamesmckinna** [comment]
So... having just completed (to my satisfaction at least) pr #1910  (cleaning up the mess of #1900 ) it occurs to me that... an R-module for commutative R should (?) be an R-Rop bimodule? I was surprised, but perhaps shouldn't be, that Rop for *commutative* R is *not* a no-op...

... and inserting the op above would ensure that your condition would hold *definitionally* ?

or have I (once again) overlooked something obvious?

**Comment by Taneb** [comment]
@jamesmckinna I believe the counterexample I brought up back in https://github.com/agda/agda-stdlib/issues/1888#issuecomment-1362920174 would still apply here. The ring of Gaussian integers is commutative. We can define a `ℤ[i]-ℤ[i]op` bimodule where the left and right multiplications are conjugate to each other rather than equal.

**Comment by jamesmckinna** [comment]
@Taneb yes, of course... what I had in mind was that by taking R-Rop as the R-S bimodule structure, your axiom about left- and right- multiplication being identified would be realisable by ... `refl` (and that that seems to be the only 'reasonable choice? I think I'd hate to have to make it `Setoid` equality of `Op₂ A` functions, by way of contrast...?)

**Comment by MatthewDaggitt** [comment]
@Taneb was this entirely fixed by https://github.com/agda/agda-stdlib/pull/1898?

**Comment by Taneb** [comment]
Yes, it is fixed
